### ISSUE TITLE
I can't use escape character \ for weird characters in folder names

### ISSUE THREAD

**Issue by RayFes** [docker/cli#1480]
**Description**

I have a folder name I want to use in a container, containing a comma. I can't use \ as escape character because that gets escaped itself automatically. 

**Steps to reproduce the issue:**
Command using --mount:
```
docker run -ti -p8888:8888 --mount type=bind, source='/run/user/7300/gvfs/smb-share:server=srvr.company.com\,share=datafldr',target=datatargetfldr myimage:latest
```
Error:
```
invalid argument "type=bind,source=/run/user/7300/gfs/smb-share:server=srvr.company.com\\,share=datafldr,target=datatargetfldr" for "--mount" flag: unexpected key 'share' in 'share=tstb'
```
*mind the double \\ in the error
Command using -v:
```
docker run -ti -p8888:8888 -v '/run/user/7300/gvfs/smb-share:server=srvr.company.com\,share=datafldr':datatargetfldr myimage:latest
```
Error:
```
docker: Error response from daemon: invalid mode: datatargetfldr
```

**Additional information you deem important (e.g. issue happens only occasionally):**
All variations of quotes, double quotes I tried give the same response. Soft links are not accepted by docker run.

**Output of `docker version`:**

```
Docker version 18.06.1-ce, build e68fc7a
```

**Output of `docker info`:**

```
Containers: 7
 Running: 0
 Paused: 0
 Stopped: 7
Images: 21
Server Version: 18.06.1-ce
Storage Driver: overlay2
 Backing Filesystem: xfs
 Supports d_type: true
 Native Overlay Diff: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host macvlan null overlay
 Log: awslogs fluentd gcplogs gelf journald json-file logentries splunk syslog
Swarm: active
 NodeID: vhbp0lazbdwgpfhwgf1m7idx5
 Is Manager: true
 ClusterID: inaoma5573wvcl1wvu2m8scms
 Managers: 1
 Nodes: 1
 Orchestration:
  Task History Retention Limit: 5
 Raft:
  Snapshot Interval: 10000
  Number of Old Snapshots to Retain: 0
  Heartbeat Tick: 1
  Election Tick: 10
 Dispatcher:
  Heartbeat Period: 5 seconds
 CA Configuration:
  Expiry Duration: 3 months
  Force Rotate: 0
 Autolock Managers: false
 Root Rotation In Progress: false
 Node Address: 172.22.199.150
 Manager Addresses:
  172.22.199.150:2377
Runtimes: runc nvidia
Default Runtime: runc
Init Binary: docker-init
containerd version: 468a545b9edcd5932818eb9de8e72413e616e86e
runc version: 69663f0bd4b60df09991c08812a60108003fa340
init version: fec3683
Security Options:
 seccomp
  Profile: default
Kernel Version: 3.10.0-862.11.6.el7.x86_64
Operating System: CentOS Linux 7 (Core)
OSType: linux
Architecture: x86_64
CPUs: 4
Total Memory: 11.56GiB
Name: ril-r0233343
ID: RLZ5:T3R4:4UDY:YX7J:OCX5:JGRM:YI3N:UTRA:MKG5:JOOS:6QME:XCQB
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Username: rayfes
Registry: https://index.docker.io/v1/
Labels:
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**
cat /etc/centos-release
CentOS Linux release 7.5.1804 (Core)

**Background:**
Trying to connect my container to a windows server with 45 TB of video data. I don't have sudo rights. sambaclient doesn't fulfill my needs. Using fuse: gvfs-mount produces these weird looking mount points. There are no options in gvfs-mount to change those.

**Comment by thaJeztah** [comment]
Yes, the shorthand `-v <source>:<target>[:<options>]` notation is known to be problematic if a path contains a colon.

In this case, you could try using the `--mount` flag, which uses a CSV syntax, and should support this (although it's still a bit fiddly, because you'd have to prevent your shell from stripping the quotes);

I'm not on a Windows machine, and of course don't have access to that SMB server, but something like this should keep the path intact;

```
docker run -d \
--mount=type=bind,\"src=/run/user/7300/gvfs/smb-share:server=srvr.company.com,share=datafldr\",target=/datatargetfldr \
nginx:alpine
```

_However_, if you want to mount smb / cifs shares, this discussion may also be of interest to you; 
https://github.com/moby/moby/issues/37161#issuecomment-392511903

**Comment by RayFes** [comment]
Bull's eye escaping the double quotes! Thanks. As such this question can be closed. 

Your _However_ is fully warranted though - this consequential question should perhaps be a new item, if you prefer me to post a new one, I will do so - .

The error I get with the docker run command is 
docker: Error response from daemon: invalid mount config for type "bind": bind mount source path does not exist. 
I am able to get till /var/run/user/7300/ folder. Just the gvfs folder or further gets me errors.

I found https://askubuntu.com/questions/401454/how-to-get-sudo-access-to-shares-mounted-by-gigolo explains other users not having access to mounts made by my userid. I guess the docker run command doesnot use my userid to approach these mounts.

Looking at your moby link-comment, I tried creating a Docker volume first specifying my login credentials and using the resulting volumename in the docker run. That either results in a "no such device" when type=volume or a "mount path must be absolute" when type=bind or using -v style. This the same for a docker volume defined with the full samba share under gvfs as well as for a docker volume pointing to the regular local folder (pythonfiles) that does work well in the container when specified directly in the run command.
Is my docker volume create incorrect? Alternatively Is there a way to specify username and password in docker run --mount? Your help is appreciated.
```
docker volume create --opt type=local --opt device=/home/m123456/Pythonfiles  \
  --opt o=file_mode=07777,dir_mode=077 VolPythonfiles
docker run -it --mount src=VolPythonfiles,dst=/pythonfils myimage:latest
docker: Error response from daemon: error while mounting volume '': error while mounting volume 
with options: type='local' device='/home/m123456/Pythonfiles' o='file_mode=07777,dir_mode=077': 
no such device.
```

**Comment by RayFes** [comment]
Sorry - wrong button #newbie

**Comment by thaJeztah** [comment]
Hm.. not sure if credentials can be passed there. Basically, the `o=` options are the options that are passed to the linux "bind" command, so whatever is supported there can be passed

**Comment by thaJeztah** [comment]
er.. s/bind/mount/
